### PR TITLE
Feat/Add trips to google calendar.

### LINKED ISSUE #247: Add trips to google calendar
This issue aims to introduce a new feature which allows a user to have the option to add their trip to their google calendar, with pre-filled information.

### PR DESCRIPTION
**Pull request by 1andreea26**
## Summary

This PR introduces the possibility to add a trip to google calendar by pressing the icon button at the top right of a trip and selecting "Add to calendar".  It opens an intent and redirects the user to the calendar app, where they  have pre-filled fields such as the name of the trip, description and start and end dates. Then, the user has the possibility to add the trip to their calendar or not.
Moreover, this feature comes with permission support for the calendar app, which if denied shows a toast to the user, and if accepted, redirects the user to the calendar app.
It also adds documentation for the Overview Screen and replacement of hardcoded strings with string resources.

## Changes

- **Screens/UI:** A new DropDownMenuItem has been introduced to the DropDownMenu in the OverviewScreen.
- **Bug Fixes/Improvements:** Addresses #247.

## Checklist

- [X] This PR resolves #247 .
- [X] Tests have been added for new functionality.
- [X] Screenshots or UI changes have been attached.
- [X] Documentation has been updated.

##  Testing

- **Manual Testing:**
    - [X] Tested on emulator / physical device.
- **Unit Tests**
    - [X] Added units tests for this

##  Related Issues/Tickets

Closes #247.

## Screenshots (if applicable)

https://github.com/user-attachments/assets/919ac327-a068-4cf3-986c-90b28ea8dc80

Screenshot 2024-12-05 at 22 01 14

## 💡 Additional Context

While I managed to test the intent and the DropDownMenuItem, permissions are hard to test.

### CODE REVIEW

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt [hunk 1]
```diff
@@ -87,7 +105,6 @@ fun OverviewScreen(
   val context = LocalContext.current
   val isConnected = status === ConnectionState.Available
 
-  Log.e("RECOMPOSE", "OverviewScreen recomposed")
   LaunchedEffect(isLoadingUser, isLoadingTrip) { isLoading = isLoadingUser || isLoadingTrip }
   LaunchedEffect(trips) {
     if (trips.isNotEmpty()) {
```
Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:110]: While the functionality is correct, the code uses multiple LaunchedEffect blocks to handle state changes, making it harder to track loading logic. Consider consolidating these side effects into a single LaunchedEffect that reacts to all relevant states (e.g., isLoadingUser, isLoadingTrip, and trips). For example, you could rely on a single ViewModel-aggregated isLoading state and respond to trips availability in one place. This would simplify the composable’s logic and enhance maintainability by providing a central point for handling state changes instead of spreading logic across multiple LaunchedEffect calls.

Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:110]: While the functionality is correct, the code uses multiple LaunchedEffect blocks to handle state changes, making it harder to track loading logic. Consider consolidating these side effects into a single LaunchedEffect that reacts to all relevant states (e.g., isLoadingUser, isLoadingTrip, and trips). For example, you could rely on a single ViewModel-aggregated isLoading state and respond to trips availability in one place. This would simplify the composable’s logic and enhance maintainability by providing a central point for handling state changes instead of spreading logic across multiple LaunchedEffect calls.

Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:110]: While the functionality is correct, the code uses multiple LaunchedEffect blocks to handle state changes, making it harder to track loading logic. Consider consolidating these side effects into a single LaunchedEffect that reacts to all relevant states (isLoadingUser, isLoadingTrip, and trips). For example, you could rely on a single ViewModel isLoading state and respond to trips availability in one place. This would simplify the composable’s logic and enhance maintainability by providing a central point for handling state changes instead of spreading logic across multiple LaunchedEffect calls. But since the functionality is correct I leave the decision up to you.

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt [hunk 2]
```diff
@@ -338,17 +412,39 @@ fun DisplayParticipants(trip: Trip, userViewModel: UserViewModel) {
...
+      Intent(Intent.ACTION_INSERT)
+          .setData(CalendarContract.Events.CONTENT_URI)
+          .putExtra(CalendarContract.Events.TITLE, trip.name)
+          .putExtra(CalendarContract.Events.DESCRIPTION, trip.description)
+          .putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, trip.startDate.toDate().time)
+          .putExtra(CalendarContract.EXTRA_EVENT_END_TIME, trip.endDate.toDate().time)
+          .putExtra(CalendarContract.Events.EVENT_TIMEZONE, TimeZone.getDefault().id)
+  // Start Activity and open calendar
+  context.startActivity(intent)
+}
```
Comment (BogdanCDev) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:450]: When launching the calendar intent, the code assumes that a calendar-capable app is available. Consider checking for a suitable activity using, for example, resolveActivity() before calling startActivity(intent). If none is found, you should handle the situation—perhaps by showing a Toast message to inform the user that no compatible application is installed. Additionally, wrapping the startActivity call in a try/catch block to handle ActivityNotFoundException is a good safety measure.

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:450]: Also, openGoogleCalendar() uses trip.startDate and trip.endDate directly without verifying their validity. Since they are sourced from user input or a network call, they might be null or invalid. To prevent crashes or unexpected behavior, ensure that Trip objects always contain valid timestamps or provide a fallback when data is missing.

**Comment by BogdanCDev** [comment]
Thank you for this solid implementation! The documentation you’ve provided is clear and helpful, making the code much easier to understand and maintain. It’s also great to see that you’re using stringResource() instead of hard-coded strings. Additionally, congratulations on achieving the target test coverage on these changes. Great work!

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt (hunk 1 above)
Comment (1andreea26) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:110]: Yes, but I think because isLoadingUser and isLoadingTrip are coming from different View Models, it would be better to leave them like this. Else, since this is not specific for this PR, I think we could address this in the last week when we look over how we could better modularise and improve our  code. Let me know if you agree!

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt (hunk 2 above)
Comment (1andreea26) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: > When launching the calendar intent, the code assumes that a calendar-capable app is available. Consider checking for a suitable activity using, for example, resolveActivity() before calling startActivity(intent). If none is found, you should handle the situation—perhaps by showing a Toast message to inform the user that no compatible application is installed. Additionally, wrapping the startActivity call in a try/catch block to handle ActivityNotFoundException is a good safety measure.

Thanks for the suggestion! I think you're right, I just handled the exception. I didn't initially add this because the android comes with a calendar downloaded so I thought it unlikely for a user to disable the app and then try to add the trip to the calendar, but handling it solves the issue more gracefully and prevents our app from crashing.

Comment (1andreea26) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: > Also, openGoogleCalendar() uses trip.startDate and trip.endDate directly without verifying their validity. Since they are sourced from user input or a network call, they might be null or invalid. To prevent crashes or unexpected behavior, ensure that Trip objects always contain valid timestamps or provide a fallback when data is missing.

Thanks! I see your point, but when you add a trip to the calendar, you're selecting it directly from overview (which means it already has startDate and endDate as valid timestamps, else it couldn't have been created). However, I understand being on the safe side, so if you think it's really necessary I can add it. Let me know what you think!

**Comment by 1andreea26** [comment]
> Thank you for this solid implementation! The documentation you’ve provided is clear and helpful, making the code much easier to understand and maintain. It’s also great to see that you’re using stringResource() instead of hard-coded strings. Additionally, congratulations on achieving the target test coverage on these changes. Great work!

Thanks for the quick review Bogdan!

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt [hunk 3]
```diff
@@ -338,17 +413,45 @@ fun DisplayParticipants(trip: Trip, userViewModel: UserViewModel) {
...
+/**
+ * Opens the default calendar app on the device with pre-filled details for a new event.
+ *
+ * This method creates an `Intent` to insert a calendar event using the trip's details such as name,
+ * description, start time, and end time.
+ *
+ * @param context The context used to launch the intent.
+ * @param trip The trip data used to populate the calendar event details.
+ */
+internal fun openGoogleCalendar(context: Context, trip: Trip) {
```
Comment (eleonoraka) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:439]: Thank you for adding the documentation for this function! However, I was wondering what's the supposed behavior when no calendar app is found or when there's an error

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt (hunk 1 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:111]: I see your perspective about the isLoadingUser and isLoadingTrip states coming from different ViewModels, which makes it logical to separate concerns. Given this, maintaining separate LaunchedEffect blocks for now ensures clarity in handling their respective states without merging responsibilities across ViewModels. Addressing the modularisation and improvement in the final review week sounds like a good plan. For now, since the functionality is correct and manageable, I agree with leaving it as is.

**Review by eleonoraka — CHANGES_REQUESTED** [review]
Great job for implementing the add trip to calendar feature! The code is clean, with documentation and the tests are complete. 
I have some suggestions:
- You could add more tests for edge cases like for null/empty trip description or location or for calendar timezone handling
- What about the activities planned during the trip? They do not appear in the calendar, could you add this?

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: > > When launching the calendar intent, the code assumes that a calendar-capable app is available. Consider checking for a suitable activity using, for example, resolveActivity() before calling startActivity(intent). If none is found, you should handle the situation—perhaps by showing a Toast message to inform the user that no compatible application is installed. Additionally, wrapping the startActivity call in a try/catch block to handle ActivityNotFoundException is a good safety measure.
> 
> Thanks for the suggestion! I think you're right, I just handled the exception. I didn't initially add this because the android comes with a calendar downloaded so I thought it unlikely for a user to disable the app and then try to add the trip to the calendar, but handling it solves the issue more gracefully and prevents our app from crashing.

Thank you for considering the suggestion about handling exceptions and checking for compatible apps.

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: > > Also, openGoogleCalendar() uses trip.startDate and trip.endDate directly without verifying their validity. Since they are sourced from user input or a network call, they might be null or invalid. To prevent crashes or unexpected behavior, ensure that Trip objects always contain valid timestamps or provide a fallback when data is missing.
> 
> Thanks! I see your point, but when you add a trip to the calendar, you're selecting it directly from overview (which means it already has startDate and endDate as valid timestamps, else it couldn't have been created). However, I understand being on the safe side, so if you think it's really necessary I can add it. Let me know what you think!

Since the creation process guarantees valid timestamps, then this might indeed be redundant. However, ensuring safety through validation could still be worthwhile, especially if there’s ever a future change in the workflow that could introduce variability. It might not be critical right now, but adding it as a safeguard could prevent issues down the line. That said, I trust your judgment here, so feel free to decide!

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt (hunk 3 above)
Comment (1andreea26) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:439]: I've just solved this issue, it's in my latest commit. Now when there's an error/no calendar app is found, selecting the menu item just displays a toast saying there's no compatible calendar app.

**Comment by 1andreea26** [comment]
> Great job for implementing the add trip to calendar feature! The code is clean, with documentation and the tests are complete. I have some suggestions:
> 
> * You could add more tests for edge cases like for null/empty trip description or location or for calendar timezone handling
> * What about the activities planned during the trip? They do not appear in the calendar, could you add this?

Thanks for the review! Regarding the edge cases, as I was just discussing with Bogdan, it doesn't quite make sense to handle cases of an empty description or location since, if the description/location are empty, they are set by default as empty strings, so when the calendar app opens, they're just empty fields which can be completed by the user. It would be counter-intuitive to introduce a text saying "No description" since the user can input a description in the calendar app.

**Comment by 1andreea26** [comment]
> * What about the activities planned during the trip? They do not appear in the calendar, could you add this?

Thanks for the suggestion, but I don't quite see where this fits in this feature. Since the feature just opens an intent, where should the activities be added? This feature doesn't directly add the trip to the calendar (in that case yes, I think activities could have been added as well), but just pre-fills some fields for the user in the calendar app (there's no activities field in the calendar app). What I could do, although I think it would be in another PR, would be to introduce the possibility to add an activity by itself to the calendar (just as for the trip). Let me know if you think that would be a cool feature.

**Comment by eleonoraka** [comment]
> Thanks for the suggestion, but I don't quite see where this fits in this feature. Since the feature just opens an intent, where should the activities be added? This feature doesn't directly add the trip to the calendar (in that case yes, I think activities could have been added as well), but just pre-fills some fields for the user in the calendar app (there's no activities field in the calendar app). What I could do, although I think it would be in another PR, would be to introduce the possibility to add an activity by itself to the calendar (just as for the trip). Let me know if you think that would be a cool feature.

Ah, you're right about the intent limitations! Though as a user, I would expect activities to be added when adding a trip to the calendar. We could add this in this PR by creating separate calendar events for each activity, or track it as a separate feature if you prefer. Let me know what you think would work better for you!

**Comment by eleonoraka** [comment]
> > Great job for implementing the add trip to calendar feature! The code is clean, with documentation and the tests are complete. I have some suggestions:
> > 
> > * You could add more tests for edge cases like for null/empty trip description or location or for calendar timezone handling
> > * What about the activities planned during the trip? They do not appear in the calendar, could you add this?
> 
> Thanks for the review! Regarding testing for edge cases, as I was just discussing with Bogdan, it doesn't quite make sense to handle cases of an empty description or location since, if the description/location are empty, they are set by default as empty strings, so when the calendar app opens, they're just empty fields which can be completed by the user. It would be counter-intuitive to introduce a text saying "No description" since the user can input a description in the calendar app. Could you be more specific with the tests you have in mind?

I initially thought we needed timezone-specific tests, but looking at your code more closely, I see they'd be redundant. 
Your implementation already handles timezones well by passing the device's timezone to the calendar intent and your existing test covers this nicely. The calendar app itself handles all the timezone conversions automatically.
I was also thinking about the "no calendar app" case. Since we can't test Toast messages directly, we could check that your code handles this scenario by making sure it catches the ActivityNotFoundException. What do you think?

**Comment by 1andreea26** [comment]
> > Thanks for the suggestion, but I don't quite see where this fits in this feature. Since the feature just opens an intent, where should the activities be added? This feature doesn't directly add the trip to the calendar (in that case yes, I think activities could have been added as well), but just pre-fills some fields for the user in the calendar app (there's no activities field in the calendar app). What I could do, although I think it would be in another PR, would be to introduce the possibility to add an activity by itself to the calendar (just as for the trip). Let me know if you think that would be a cool feature.
> 
> Ah, you're right about the intent limitations! Though as a user, I would expect activities to be added when adding a trip to the calendar. We could add this in this PR by creating separate calendar events for each activity, or track it as a separate feature if you prefer. Let me know what you think would work better for you!

By creating separate calendar events for each activity, you mean creating the possibility for the user to add an activity to the calendar, right? If this is the case then yes, that could be done though I would not add it in this PR as it's a different feature (despite the similar implementation). Let me know what you mean more specifically!

**Comment by 1andreea26** [comment]
> > > Great job for implementing the add trip to calendar feature! The code is clean, with documentation and the tests are complete. I have some suggestions:
> > > 
> > > * You could add more tests for edge cases like for null/empty trip description or location or for calendar timezone handling
> > > * What about the activities planned during the trip? They do not appear in the calendar, could you add this?
> > 
> > 
> > Thanks for the review! Regarding testing for edge cases, as I was just discussing with Bogdan, it doesn't quite make sense to handle cases of an empty description or location since, if the description/location are empty, they are set by default as empty strings, so when the calendar app opens, they're just empty fields which can be completed by the user. It would be counter-intuitive to introduce a text saying "No description" since the user can input a description in the calendar app. Could you be more specific with the tests you have in mind?
> 
> I initially thought we needed timezone-specific tests, but looking at your code more closely, I see they'd be redundant. Your implementation already handles timezones well by passing the device's timezone to the calendar intent and your existing test covers this nicely. The calendar app itself handles all the timezone conversions automatically. I was also thinking about the "no calendar app" case. Since we can't test Toast messages directly, we could check that your code handles this scenario by making sure it catches the ActivityNotFoundException. What do you think?

Thanks for the suggestion! I tried to implement this, but since we can't test toast messages directly, unfortunately since the app throws a toast in case of such an exception, there is no way to test it (because the tests will throw exceptions since they are trying to access a toast), thus such a test would not pass. If you know any other way in which I could test this, let me know, but I currently haven't found a way (except removing the toast, which would not be user friendly since the app would not show anything in case of exception).

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt (hunk 2 above)
Comment (1andreea26) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: I will leave them like this because I've tried to implement them in try-catch blocks,  but this reduces our line coverage under 80% , and I can't provide tests for them (since I have no way to create a trip with an invalid startDate/endDate to make the method .toDate() or .time to fail and even if I did, I could not test because they are throwing toasts in case of an exception). As this is not critical for now, if you agree, I would leave it like the original implementation.

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: I agree that this is a good solution for now. It’s worth noting that this is not the original implementation, but the current one in the branch, which includes the modification to handle ActivityNotFoundException with the toast fallback, which I believe to be quite important. For now, maintaining the code coverage as high as possible is important, so we can keep it as is. We can revisit the other aspects in the final sprint when the app is closer to release to manage these better.

**Review by BogdanCDev — APPROVED** [review]
Thank you for taking my suggestions into account! The implementation now looks really solid, and I appreciate the effort you’ve put into it. There are a few points I previously mentioned that haven’t been implemented yet, but I understand their inclusion might decrease the code coverage at this stage. I believe we can revisit these during the final sprint when we’ll be in a better position to manage the coverage impact before the release.

That said, the current functionality is working well, and no bugs are present, so I think this is ready to merge into the main branch. Before merging, please address the Sonar security hotspot and mark it as reviewed to ensure we’re adhering to security best practices. Great work, and thank you for addressing the feedback so effectively!

**Comment by 1andreea26** [comment]
> Thank you for taking my suggestions into account! The implementation now looks really solid, and I appreciate the effort you’ve put into it. There are a few points I previously mentioned that haven’t been implemented yet, but I understand their inclusion might decrease the code coverage at this stage. I believe we can revisit these during the final sprint when we’ll be in a better position to manage the coverage impact before the release.
> 
> That said, the current functionality is working well, and no bugs are present, so I think this is ready to merge into the main branch. Before merging, please address the Sonar security hotspot and mark it as reviewed to ensure we’re adhering to security best practices. Great work, and thank you for addressing the feedback so effectively!

Thanks for the approval and help, Bogdan!

**Comment by eleonoraka** [comment]
> > > Thanks for the suggestion, but I don't quite see where this fits in this feature. Since the feature just opens an intent, where should the activities be added? This feature doesn't directly add the trip to the calendar (in that case yes, I think activities could have been added as well), but just pre-fills some fields for the user in the calendar app (there's no activities field in the calendar app). What I could do, although I think it would be in another PR, would be to introduce the possibility to add an activity by itself to the calendar (just as for the trip). Let me know if you think that would be a cool feature.
> > 
> > 
> > Ah, you're right about the intent limitations! Though as a user, I would expect activities to be added when adding a trip to the calendar. We could add this in this PR by creating separate calendar events for each activity, or track it as a separate feature if you prefer. Let me know what you think would work better for you!
> 
> By creating separate calendar events for each activity, you mean creating the possibility for the user to add an activity to the calendar, right? If this is the case then yes, that could be done though I would not add it in this PR as it's a different feature (despite the similar implementation). Let me know what you mean more specifically!

Yes, exactly! I meant giving users the ability to add individual activities as calendar events. I agree with your thinking that while it's similar in implementation, it would be better as a separate feature rather than adding it to this PR. That way we keep the current PR focused on the trip calendar integration, while this activities feature could be developed separately.

**Review by eleonoraka — APPROVED** [review]

**Comment by eleonoraka** [comment]
> Thanks for the suggestion! I tried to implement this, but since we can't test toast messages directly, unfortunately since the app throws a toast in case of such an exception, there is no way to test it (because the tests will throw exceptions since they are trying to access a toast), thus such a test would not pass. If you know any other way in which I could test this, let me know, but I currently haven't found a way (except removing the toast, which would not be user friendly since the app would not show anything in case of exception).

I didn't think about it! Let's leave it like that then. Great work!

**Comment by 1andreea26** [comment]
> > Thanks for the suggestion! I tried to implement this, but since we can't test toast messages directly, unfortunately since the app throws a toast in case of such an exception, there is no way to test it (because the tests will throw exceptions since they are trying to access a toast), thus such a test would not pass. If you know any other way in which I could test this, let me know, but I currently haven't found a way (except removing the toast, which would not be user friendly since the app would not show anything in case of exception).
> 
> I didn't think about it! Let's leave it like that then. Great work!

Thanks Eleonora!
